Clustering/Hdbscan/src/main/java/org/tribuo/clustering/hdbscan/HdbscanTrainer.java

```diff
@@ -779,31 +804,68 @@ private static List<ClusterExemplar> computeExemplars(SGDVector[] data, Map<Inte
+                                maxInnerDist));
+                    }
                 }
-
-                // First, get the entries that will be used for cluster exemplars.
-                // The first node is polled from the tree, which has the lowest outlier score out of the remaining
-                // points assigned this cluster.
-                List<Entry<Double, Integer>> partialClusterExemplars = new ArrayList<>();
-                Stream<Integer> intStream = IntStream.range(0, numExemplarsThisCluster).boxed();
-                intStream.forEach((i) -> partialClusterExemplars.add(outlierScoreIndexTree.pollFirstEntry()));
-
-                // For each of the partial exemplars in this cluster, iterate the remaining nodes in the tree to find
-                // the maximum distance between the exemplar and the members of the cluster. The other exemplars don't
-                // need to be checked here since they won't be on the fringe of the cluster.
-                for (Entry<Double, Integer> partialClusterExemplar : partialClusterExemplars) {
-                    SGDVector features = data[partialClusterExemplar.getValue()];
-                    double maxInnerDist = Double.NEGATIVE_INFINITY;
-                    for (Entry<Double, Integer> entry : outlierScoreIndexTree.entrySet()) {
-                        double distance = dist.computeDistance(features, data[entry.getValue()]);
-                        if (distance > maxInnerDist){
-                            maxInnerDist = distance;
+                // Otherwise, the number of exemplars is greater than the size of the tree, which is most likely to occur
+                // with a contrived dataset. This indicates that there were several points in this cluster with non-unique
+                // outlier scores.
+                else {
+                    int outlierScoreTreeSize = outlierScoreIndexTree.size();
+                    List<Pair<Double, Integer>> partialClusterExemplarsList = new ArrayList<>();
+                    // When many points in the cluster have non-unique outlier scores, first we'll collect the items from
+                    // the tree.
+                    Stream<Integer> intStreamTree = IntStream.range(0, outlierScoreTreeSize).boxed();
+                    intStreamTree.forEach((i) -> {
+                        Entry<Double, Integer> treeEntry = outlierScoreIndexTree.pollFirstEntry();
+                        partialClusterExemplarsList.add(new Pair<>(treeEntry.getKey(), treeEntry.getValue()));
+                    });
+
+                    // To determine the remaining exemplars, the best thing to do is randomly sample them from all the
+                    // points in this cluster. This could introduce duplicate exemplar points, but that is safer than
+                    // reducing the number of exemplars.
+                    SplittableRandom rand = new SplittableRandom(exemplarSampleSeed);
```

Comment:
How much of an edge case is this randomized fallback? We have special treatment for all other RNGs in Tribuo where they are members of the trainer and split under a lock to preserve provenance information, which ensures that repeated runs of a trainer on the same data give different answers when not controlling the RNG state (e.g. when used in an ensemble). If this is purely an edge case issue then it might be ok to leave it as is, but if it might occur relatively frequently then it's probably worth using the same idiom we do elsewhere.

Reply:
This edge case seems to be quite rare, but if there is a "best solution" which can be implemented here we should strive towards that. I'll assume that the implementation of this special treatment demonstrated in the `KMeansTrainer` is a good example to follow.

Reply:
Yeah, this synchronized idiom along with storing an rng as a field - https://github.com/oracle/tribuo/blob/main/Clustering/KMeans/src/main/java/org/tribuo/clustering/kmeans/KMeansTrainer.java#L278.

Reply:
I've added a commit which adds a RNG as a field, and the supporting logic. Let me know if anything else might be needed.